src/pterms/PtStore.cc

```diff
@@ -67,49 +99,91 @@ SymRef PtStore::lookupSymbol(const char* s, const vec<PTRef>& args) {
+                        candidates.push(ctr);
+                    }
+                } else if (t.nargs() < args.size_() && t.pairwise()) {
                     int j = 0;
                     for (; j < args.size(); j++) {
                         SymRef argt = pta[args[j]].symb();
                         if (symstore[argt].rsort() != t[0]) break;
                     }
-                    if (j == args.size()) return ctr;
+                    if (j == args.size()) {
+                        candidates.push(ctr);
+                    }
                 }
             }
-        }
-
-        for (int i = 0; i < trefs.size(); i++) {
-            SymRef ctr = trefs[i];
-            const Symbol& t = symstore[ctr];
-            if (t.nargs() == args.size_()) {
-                // t is a potential match.  Check that arguments match
-                uint32_t j = 0;
-                for (; j < t.nargs(); j++) {
-                    SymRef argt = pta[args[j]].symb();
-                    if (t[j] != symstore[argt].rsort()) break;
+        } else {
+            for (int i = 0; i < trefs.size(); i++) {
+                SymRef ctr = trefs[i];
+                const Symbol &t = symstore[ctr];
+                if (t.nargs() == args.size_()) {
+                    // t is a potential match.  Check that arguments match
+                    uint32_t j = 0;
+                    for (; j < t.nargs(); j++) {
+                        SymRef argt = pta[args[j]].symb();
+                        if (t[j] != symstore[argt].rsort()) break;
+                    }
+                    if (j == t.nargs()) {
+                        candidates.push(ctr);
+                    }
                 }
-                if (j == t.nargs())
-                    return ctr;
             }
         }
+    }
+    if (candidates.size() == 0) {
+        return SymRef_Undef; // Not found
+    } else if (candidates.size() == 1 and sort == SRef_Undef) {
+        return candidates[0];
+    } else if (candidates.size() == 1 and sort != SRef_Undef) {
+        if (symstore[candidates[0]].rsort() != sort) {
+            return SymRef_Undef;
+        } else {
+            return candidates[0];
+        }
+    }
+
+    assert(candidates.size() > 1);
+    if (sort == SRef_Undef) {
+        throw OsmtApiException("Ambiguous symbol: `" + std::string(s) + "'");
+    }
+
+    assert(sort != SRef_Undef);
+
+    int i, j;
+    for (i = j = 0; i < candidates.size(); i++) {
+        if (symstore[candidates[i]].rsort() == sort) {
+            candidates[j++] = candidates[i];
+        }
+        if (j > 1) {
+            throw OsmtInternalException("System has > 1 symbols with same argument and return sorts");
+        }
+    }
+    candidates.shrink(i-j);
 
+    if (candidates.size() == 0) {
+        return SymRef_Undef;
     }
-    // Not found
-    return SymRef_Undef;
+    assert(candidates.size() == 1);
+    assert(symstore[candidates[0]].rsort() == sort);
+    return candidates[0];
```

Comment:
This looks OK, although, I would still prefer to use an STL algorithm for this.

Reply:
Yes, maybe so..  But at least with std::remove_if the exception throwing would mean that the predicate would have to maintain a state, which takes away at least some of its appeal.

Reply:
As discussed, this cannot be a performance overhead in working code.